### ISSUE TITLE
[RFC] Simplification of Security Stance and Removal of Bespoke Signing Requirements

### ISSUE THREAD

**Issue by autodidaddict** [wasmCloud/wasmCloud#1119]
# RFC: Simplification of Security Stance and Removal of Bespoke Signing Requirements

## Introduction

wasmCloud, and by extension, the lattice, is designed to operate in a zero trust environment. The workloads that are scheduled are `.wasm` files that are inherently untrusted. Over time, we've learned from various architecture changes and evolutions that many of the security measures originally put in place are providing little to no security while introducing large amounts of complexity and poor developer experience. This RFC outlines a way to gather up all of the lessons learned from wasmCloud security over the years to streamline the security posture while also removing the only remaining platform lock-in feature - the custom JWT section.

## Motivation

A lot of the decisions we've made with regard to security are based on assumptions that are no longer valid. Further, some of the changes we've made to increase the prominence and "stickiness" of things like link definitions have given us chances to smooth out the developer experience that we haven't yet exploited.

## Security Review
The following is a summary of some thoughts on the current state of things viewed through the lens of hindsight and lessons learned.

* **Invocation signing** - the original thought process behind this was in service of defense in depth. The idea was that if someone was able to gain access to the NATS infrastructure with sufficient privilege to access the control topics, they could forge invocations to actors or to providers. The authN version of this is that a signed invocation means that it has a provable and verifiable source (the issuer). The issuer of an invocation is a host within the lattice and so this leads to the trusted issuer list within a lattice. The main driver behind the trusted issuer list was to prevent dev/stage/test workloads from being able to run in production. This feels like overkill.  Securing access to the `rpc` and `ctl` subject namespaces should be sufficient. In today's design, invocations are untrusted and are only allowed upon validating the JWT.  The sender of an invocation doesn't currently validate the receiver. I honestly think just using NATS security to restrict access to these topics is good enough because some of the other countermeasures like invocation signing are, in hindsight, security theater.
* **Actor to Actor Calls** - in the original design, actors are required to have signed claims that grants them access to providers of a given contract. This is why they're specifically signed with access to things like `wasmcloud:messaging`. Even if the code were to make an attempt to call a linked messaging provider, it would be rejected without the appropriate claims. This is why we also include claims in the signed part of a JWT, allowing the receiving host to validate that the invocation from the sender is actually allowed. The lattice design is completely distributed, so nothing ever assumes locality, so remote invocations are, as mentioned in 1, _untrusted_. There is actually an [ADR](https://github.com/wasmCloud/wasmCloud/blob/main/adr/0006-actor-to-actor.md) (and possibly a blog post) on why we chose to allow actor-to-actor calls by default. The main reason was that this type of call cannot be used as a privilege escalation vector, so it's not inherently dangerous to allow it. We developed the policy service protocol to allow for enterprises to enforce stricter policies on which actors can invoke which other actors.
* **Call alias** - the reason call alias is burned into the module as part of the claims was so that an actor could be invoked by alias. We used the alias in demos like ping/pong. the aliases are there so that actors can communicate with each other even when their public key changes. Going back to item 1 - if an actor is issued by two different authorities, one for dev and one for prod, then it's possible that its public key will change between dev and prod. The call alias is a developer-friendly handle that _does not change between environments or actor seed keys_.

### Motivation Thoughts
Thinking back over the years of wasmCloud's evolution, it's easy to see how we got here, _but it's too complex_. The link definition is something that can only be supplied by operations, not by the developer who wrote the code. That link definition can also only ever exist if the publisher had privileged access to the right topic(s). So it's dramatically simpler to just assume that if a link exists between any two entities (actor or provider), then those two entities are allowed to communicate. This still sticks with the spirit of untrusted computing, but makes the _link an assertion of privilege and not signed claims_. So the entire invocation privilege checking could be simplified all the way down to "does a link exist between caller and callee? if so, the call is allowed". 

This would apply to both providers and actors. If wasmCloud supported a link type that was between two actors, then you wouldn't need policy services to determine whether an actor call can take place.

## Detailed Design
There are a number of things that would change in wasmCloud to support this, and the changes would be wide-sweeping and touch nearly all aspects of the runtime. However, the vast majority of changes would be in the removal of code, removal of complexity, and removal of bespoke tooling that only applies to the wasmCloud environment. In short, it makes wasmCloud more ready to embrace the component model, wit, and "off the shelf" standards, _without sacrificing security_.

The following is a high level list of what would change:
* wasmCloud modules (`.wasm`) and components would no longer require signing _of any kind_. The custom `JWT` section embedded into wasmCloud modules would no longer be required. There are other things that can be done with wit and components and standard tooling to add useful metadata to a component.
* Call verification - these changes would take place to invocation security:
    * Invocations no longer signed
    * If a link exists between two entities, those two entities are allowed to communicate. This applies to actors and providers. If a link does not exist, those entities _cannot_ communicate. 
    * Policy service would no longer be consulted to determine if 2 entities can communicate
* Link definitions would support actor-to-actor links and support a `target_alias` field on actor-to-actor links to allow call aliases to continue to work without signed JWTs. 

### Actor Identity
So how do we uniquely identify actors if we aren't signing their `.wasm` files with a JWT embedded in a custom section? If we concede that we don't need to block against actor identity forgery because comms can only ever take place over valid link definitions (which are a secured entity only creatable by those with control interface access), then we don't need a cryptographic public key. 

Instead, when we make a control interface to request the scaling of an actor, we can supply the unique ID to be associated with that actor. We could also use the OCI reference URL or some combination of the two as the established identity. The net result is that the identity comes from the launch/scale request, and not from the raw binary.

So now an actor might be visible to tooling, APIs, and UIs as `mega-app-3000` rather than `OD7II4B57ENTOV3JJYTOA3FDK6XCUXOOFXJNUVCHOW4JHAGXWHMXBZQU`.

There's a potential footgun there, but it's still pretty small (accidentally using the wrong name for a given component). The consequences of this would be that the incorrectly named/identified module would end up as part of a different queue group than the other instances of the same module.

## Backwards Compatibility

Since we'd be validating less and checking less things, old modules previously signed using a custom section would continue to work. The only think that would need to happen is that links would need to be added between actors that previously enjoyed unfettered communication. Extra signatures and things on invocations and components could be ignored by the new version.

## Alternatives Considered

The only alternatives considered were making some of the intended changes in small pieces rather than part of a single effort. We could also do nothing and leave signing in place for actors and invocations.

## Conclusion

In conclusion, I think this change would make for an absolutely game-changing simplification across the wasmCloud ecosystem, getting it ready for the kind of developer experience one would expect from a 1.0 product. This entire RFC could be described as "stop signing modules and invocations" and reap the benefits of the new lack of complexity and increased resonance with public standards. :thinking: Perhaps there is such a thing as being too paranoid when it comes to security.

**Comment by rvolosatovs** [comment]
Thank you for writing this up @autodidaddict!

Totally agree with the RFC. Taking this a little step further though, since we now would have links between actors, do you think there is still a reason to differentiate between a "call alias" and a "link name"?

**Comment by autodidaddict** [comment]
Oh that's a good idea. You could use the link name instead of the call alias, since the actor would need compile-time knowledge of either of those

**Comment by vados-cosmonic** [comment]
Do we lose anything by removing the signing? While I wholeheartedly agree with the improvement in UX, it seems like we *might* be losing a few things:

- losing a kind of identity/provenance indicator for actors
- losing a method of secure communication for actors (i.e. some asymmetric + symmetric encryption scheme for messages)
- increasing our coupling to the NATS ecosystem into NATS even more

**Comment by autodidaddict** [comment]
* provenance - agreed, that would go away. I think the lingering question is, do we need it? It feels like provenance might be something the at-large wit/wasm community might want to solve in a standard way
* encryption - if you want to encrypt data from one sender to one specific target, you could use xkeys and still not have to do signing
* I would disagree with coupling. Nothing about this proposal makes us more reliant on NATS. it does, however, accept that NATS security is a boundary on which we depend. But, that's basically a statement every NATS user makes.

**Comment by brooksmtownsend** [comment]
There's a lot in this RFC! I want to break down some of my thoughts about each point, so bear with the wall

1. Invocation Signing
    a. I don't think this involves a massive amount of complexity, minus the immediate developer iteration loop case where you schedule a component and then want to `wash call` a specific operation. I would rather combat this issue by signing and storing a cluster key like we do with issuer/module keys and then using that automatically with wash call, making it opaque to the user. The main reason for this is that I still feel this is a valuable level of defense in depth, if the RPC connection is compromised in a production environment then you are vulnerable to traffic snooping & traffic replays, but not arbitrary injected invocations.
3. Actor <-> Actor and Actor <-> Provider Links
    a. I agree that we should be able to drop claims signing from the component. Looking to the future, inspecting the embedded wit metadata should tell us everything we need to know about what the component needs to access, and linking at runtime is enough of an assertion that the component can talk to another component (actor or provider.)
    b. I like the idea of linking actor to actor, especially with the notion that a link (shall we call it binding?) is just a runtime connection between two unique references without configuration and we can handle configuration separately. This feeds somewhat into the security discussion from yesterday's meeting.
    c. I'm very into the idea of linking between unique references that are synthesized from the registry reference for the actor/provider. This gets us much closer to implementing #363 as long as we can iron out possible edge cases from local development and the difference between an OCI reference that's immutable (blah.cr.io/echo:0.3.4) vs something that could change (bloh.cr.io/echo:latest).
    d. Links between actor and provider are inherently bidirectional, do we need to handle the case where one actor always just calls the other?
4. Wasm signing
    a. I agree that this should be something solved by the broader community, and I think our implementation is really solid. The use of offline-verifiable ed25519 keys that are "human readable" and simple to understand is very nice. I think we should try and propose this strategy for the broader Wasm signing story, as it has the potential to be the standard with how flexible it is.
    b. All that being said, I would be in favor of making the verification of signed modules an optional measure that can be disabled at runtime. This would support quick iteration in the `wash up` or `wash run` case, but still enforce a production level of security that no unsigned modules can run.
5. Call alias
    a. With the proposed above, I don't see too many changes other than maybe declaring call aliases as link names?

Lots of statements and questions and assumptions above. Overall I'm a huge fan of the idea of simplifying the steps we've taken for production grade security and enabling a better out-of-the-box experience in places where the security is only giving us the warm fuzzies instead of preventing an actual attack scenario.

**Comment by connorsmith256** [comment]
Brooks beat me to it but said everything I was going to. Summarizing for my own edification:

- I think invocation signing should stay, but can be disabled in a dev context (`wash up`) (like enabling starting components from a file instead of published artifact). I don't see how this is security theater, nor do I see it imposing a burden on developers. Today, even if my RPC connection is compromised, I can trust that no new invocations can be injected into the lattice, unless my cluster keys are _also_ compromised. Whereas if we remove invocation signing, the burden lies solely with securing the NATS connection. We also lose the ability to distinguish traffic among multiple cluster issuers. This eliminates configurations where I may have greater-than-zero trust for some hosts in a lattice, but not for all. A contrived scenario: I might be okay with an untrusted host being served traffic via an HTTP server provider on a trusted host, but might not be okay with an untrusted host using an HTTP client provider to make outbound connections from a trusted host. I can enable that today with a policy service that inspects the issuers.
- I think provenance is an important consideration, and I don't think wasmCloud should do something bespoke. However, I'm not aware of a "standard" for this today. I agree we should get the ball rolling among interested parties on what such a standard could look like; until then, I don't see what we would gain by removing our current signing approach
  - I think Brooks brought up a good point here that this is another thing that could be optionally disabled in a dev context (`wash up`). This could be useful to run off-the-shelf components in the wasmCloud ecosystem without forcing users to sign them
- I am in favor of removing embedded capability claims and allowing operators to control how components can communicate, via links 👍 
- I'm also in favor of supporting actor-to-actor links 👍 
- Mostly unrelated to this RFC, but we've been discussing "config" generally, and I think these changes help move us to a world where links are used (only) to allow runtime communication, and config is provided to actors/providers via a separate mechanism
- I'm also in favor of treating a reference (e.g. OCI URL) as the "primary key" for an actor/provider in a host's inventory, and not using the public key for this anymore 👍 

Thank you for the extensive writeup @autodidaddict, this got me thinking deeply about our security architecture in a way I hadn't in a while

**Comment by autodidaddict** [comment]
Yeah one thing I like about this is that things can be done iteratively. So you could replace "can invoke" checks with link existence and leave everything else in place and evaluate as needed

**Comment by vados-cosmonic** [comment]
> I'm also in favor of treating a reference (e.g. OCI URL) as the "primary key" for an actor/provider in a host's inventory, and not using the public key for this anymore 👍

We should probably use a prefix and allow a configurable identifier here -- opening it up to the URLs is likely going to create even more footguns than we've already identified. A simple prefix would allow us to use content hashes, different strategies in the future, and allow for graceful migration forwards.

> The biggest issue here in my opinion is the false sense of security - the invocation signing as designed today only works for a very limited set of use cases and applications and their developers absolutely should not assume that it makes their deployment secure. NATS security is what makes wasmCloud deployments secure, compromised NATS means fully compromised application today. Only once NATS security is in place users should start thinking about things like invocation signing, it's an auxillary feature, which does very little impact to actual security of the system.

While I agree that NATS is the underpinning of security for `wasmcloud` today (so in general if it is compromised we are dramatically less likely to remain secure), invocation (i.e. request) (and response signing) *is* the solution to the problem. Our problem is that we did not go far enough (as you noticed): 

> Now the "real" ways to solve the security issues here would be:
> a. Encryption
> b. Invocation response signing

The defence-in-depth posture is useful here -- in a world where we are fully signed, an actor has to:
- Break into NATS
- Break into the host that is sending the messages to do invocation signing
- (optional, in a world where we have invocation response signing) Break into the host that is receiving invocations to influence response signing/steal secrets

IMO encryption should also be possible and easy with the wasmcloud toolkit.

> Regardless, it's really the host identity that we care about, since that's the party that really is (or should be) communicating over the lattice, not the identity of the (author of the) workload, whether that's an actor or provider (which is especially true once it's the host's responsibility to subscribe/publish to NATS topics associated with (mostly WASI) providers). Signing of WebAssembly workloads (or providers) really does not provide much value to wasmCloud.

Since today hosts spawn providers (and would even more closely manage/interact with WASI providers), then it follows that it's not really possible for an actor to hide information from a host, so it's OK that it's the host's responsibility to subscribe/publish to NATS topics on behalf of the provider.

That said, in the future I'd like to see remote providers registerable to a host. In that future, providers can run somewhere completely different, and have their *own* authorized connections to the NATS cluster, which means they would navigate their own signing/protection when sending messages. 

Taking a stance that requires the host to originate, sign and/or encrypt all messages would box us into providers not being able to ever leave the vicinity of a host and introduce latency and inflexibility. I think we should keep up the zero trust at the level *inside* the NATS connection. 

I'm still convinced that removing signing increases the reliance on NATS here (and hopefully it's more obvious now) -- in a world where we speak over some other conduit, signing at the lower level helps us make that switch without worrying about becoming completely unsecured since we expect to send everything over NATS and use it's security guarantees. This also protects us from the admittedly unlikely chance of NATS compromises.

This also has effects on the trust requirements of the setup. If providers *must* trust the host they're attached to, this is more onerous than only needing to trust hosts to relay control messages. It's the difference between the host being a proxy/gateway and the host being the only communication avenue.

**Comment by thomastaylor312** [comment]
> I think provenance is an important consideration, and I don't think wasmCloud should do something bespoke. However, I'm not aware of a "standard" for this today. I agree we should get the ball rolling among interested parties on what such a standard could look like; until then, I don't see what we would gain by removing our current signing approach

I very much agree here and let me explain why: having done a bunch of stuff in both container and Helm space, both of those made the decision to forego provenance in the beginning, added it later, and now it is a pain to use and a lot of people don't use it. Based on what was outlined here, I don't think it would be too hard to continue with what we're doing, see what the community wants to do (there could be something here around reusing prior art from the CNCF space, but that is a discussion from elsewhere), and then remove or roll into something else later. 

I could be convinced away from this btw, just explaining why I worry a bit around removing it now that we pretty much automate this away for people at this point. I definitely get it could be an easier dev experience.

**Comment by thomastaylor312** [comment]
Oh and also, thanks for the detailed write up here @autodidaddict. This was great technical reading

**Comment by autodidaddict** [comment]
I have thoughts on each of these on their own, which I think further bolsters the case for splitting them all out and working them independently.

**Comment by thomastaylor312** [comment]
Should we keep this open until we split them out as separate RFCs?

**Comment by brooksmtownsend** [comment]
Okay, phew, I believe I've broken up each of the action items above into separate RFCs as mentioned. I'm closing this issue as a way to avoid just letting it go stale and get closed anyways, but **please reopen this if I missed a discussion point above.** I mainly wanted to break out this RFC into individual topics so we can proceed on them in parallel.

cc @autodidaddict @rvolosatovs @thomastaylor312 @connorsmith256 @vados-cosmonic
